### ISSUE TITLE
None is subtype of everything?

### ISSUE THREAD

**Issue by jasonkuhrt** [facebook/pyre-check#65]
Why does `basic` type check?

```py
import random
import string


def gen(allowed_chars: string, length: int) -> str:
    ''.join([random.choice(allowed_chars) for _ in range(1, length)])


def basic(length: int) -> str:
    gen(string.ascii_letters + string.digits, length)
```

```
❯ pyre check
 ƛ Found 1 type error!
source/lib/random_string.py:6:4 Incompatible return type [7]: Expected `str` but got `None`.
```

I tried strict mode on the module to no effect (unsurprisingly since its about `Any`, not `None`).

**Comment by dark** [comment]
Hi @jasonkuhrt , Pyre is warning you that you `gen` does not return any value. You need to add a `return` statement for that to work. That is also the case for the `basic` function.
`basic` does not throw any visible error here because `string` is not a type (for the record, this is the stub for the string module: https://github.com/python/typeshed/blob/master/stdlib/3/string.pyi)
You can see more low-level errors like this with `pyre --debug check`:

> foo.py:68:4 Incompatible return type [7]: Expected `str` but got `None`.
> foo.py:70:0 Undefined type [11]: Type `string` is not defined.

After correcting the file, no errors are generated for me:

```python
import random
import string

def gen(allowed_chars: str, length: int) -> str:
    return ''.join([random.choice(allowed_chars) for _ in range(1, length)])

def basic(length: int) -> str:
    return gen(string.ascii_letters + string.digits, length)
```

Feel free to reopen this issue if you have further related questions.

**Comment by jasonkuhrt** [comment]
> basic does not throw any visible error here because string is not a type

I don't understand this. `basic` returns `None` but its type says that it will return `str`. What am I doing wrong exactly?

**Comment by dark** [comment]
Let me rephrase - the error about a missing return in `basic` is being masked by the use of an unknown type `string`. In cases like this we are trying to be conservative and not display the error because the type might be legitimate and simply not being defined in a stub.

**Comment by jasonkuhrt** [comment]
@dark I see, thanks. Could I fix this at the project level via local defs? The issue is that I'm using e.g. `ascii_letters` which has def:

```
ascii_letters = ...  # type: str
```

**Comment by dark** [comment]
@jasonkuhrt Using the imported `string` module is totally fine -- and it works!
The problem was with the following incorrect type definition for the parameter of `gen`: `allowed_chars: string`. String is a module, not a type, so the correct definition would be `allowed_chars: str`. Please see the fixed code I posted in my first comment for all things I needed to change.

**Comment by jasonkuhrt** [comment]
@dark ahhhhh sorry my eyes glazed over that, of course. Thanks.
